### PR TITLE
[CARBONDATA-4111] Filter query having invalid results after add segment to table having SI with Indexserver

### PR DESCRIPTION
**Pull request by ShreelekhyaG**
### Why is this PR needed?
When the index server is enabled, filter query on SI column after alter table add sdk segment to maintable throws `NoSuchMethodException`  and the rows added by sdk segment are not returned in the result.
 
 ### What changes were proposed in this PR?
Added segment path in index server flow, as it is used to identify external segment in filter resolver step. 
No need to load to SI, if it is an add load command.
Default constructor for `SegmentWrapperContainer` declared.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5352/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3592/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3598/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5358/

**Comment by ydvpankaj99** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/indexstore/SegmentWrapperContainer.java [hunk 1]
```diff
@@ -31,6 +31,9 @@
 
   private SegmentWrapper[] segmentWrappers;
 
+  public SegmentWrapperContainer() {
```
Comment (vikramahuja1001) [inline core/src/main/java/org/apache/carbondata/core/indexstore/SegmentWrapperContainer.java:34]: why is this change required?

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5369/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3611/

File: core/src/main/java/org/apache/carbondata/core/indexstore/SegmentWrapperContainer.java (hunk 1 above)
Comment (ShreelekhyaG) [reply core/src/main/java/org/apache/carbondata/core/indexstore/SegmentWrapperContainer.java:34]: `NoSuchMethodException` is thrown as the default instance is called from reflectionUtils.

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java [hunk 2]
```diff
@@ -221,7 +221,7 @@ public void deserializeFields(DataInput in, String[] locations, String tablePath
       indexUniqueId = in.readUTF();
     }
     String filePath = getPath();
-    if (filePath.startsWith(File.separator)) {
+    if (!FileFactory.isFileExist(filePath)) {
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java:224]: Please revert this change to do a rpc call. Use FileFactory.getupdatedFilePath instead as discussed

File: core/src/main/java/org/apache/carbondata/core/index/IndexInputFormat.java [hunk 3]
```diff
@@ -159,6 +162,19 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
           if (filterResolverIntf != null) {
             filter.setExpression(filterResolverIntf.getFilterExpression());
           }
+          for (Segment segment : segmentsToLoad) {
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/index/IndexInputFormat.java:165]: I think, after [PR-3656](https://github.com/apache/carbondata/pull/3656), we are not loading external segment to SI. External segment will be queried from main table only. But currently, i think SILoadEventListenerForFailedSegments listener is loading external segment while, si repair is enabled. Please check and handle it

Comment (ShreelekhyaG) [reply core/src/main/java/org/apache/carbondata/core/index/IndexInputFormat.java:165]: Done. Added check to skip for external segments.

File: integration/spark/src/main/java/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java [hunk 4]
```diff
@@ -51,9 +51,10 @@
   public static List<String> getListOfValidSlices(LoadMetadataDetails[] details) {
     List<String> activeSlices = new ArrayList<>(CarbonCommonConstants.DEFAULT_COLLECTION_SIZE);
     for (LoadMetadataDetails oneLoad : details) {
-      if (SegmentStatus.SUCCESS.equals(oneLoad.getSegmentStatus())
+      // No need to consider external segments for SI load
+      if (oneLoad.getPath() == null && (SegmentStatus.SUCCESS.equals(oneLoad.getSegmentStatus())
```
Comment (Indhumathi27) [inline integration/spark/src/main/java/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java:55]: Please add code to not repair si, when add load is executed

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5395/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3635/

File: integration/spark/src/main/java/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java (hunk 4 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/java/org/apache/spark/sql/secondaryindex/load/CarbonInternalLoaderUtil.java:55]: Done

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java (hunk 2 above)
Comment (ShreelekhyaG) [reply core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java]: As discussed, using FileExists only for local file path.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5398/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3638/

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java [hunk 5]
```diff
@@ -221,7 +223,13 @@ public void deserializeFields(DataInput in, String[] locations, String tablePath
       indexUniqueId = in.readUTF();
     }
     String filePath = getPath();
-    if (filePath.startsWith(File.separator)) {
+    boolean isLocalFile = FileFactory.getCarbonFile(filePath) instanceof LocalCarbonFile;
+    // If it is external segment path, table path need not be appended to filePath
+    // Example filepath: hdfs://hacluster/opt/newsegmentpath/
+    // filePath value would start with hdfs:// or s3:// . If it is local
+    // ubuntu storage, it starts with File separator, so check if given path exists or not.
+    if ((!isLocalFile && filePath.startsWith(File.separator)) || (isLocalFile && !FileFactory
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java:231]: the comment is not clear, please rewrite it with better example and scenarios

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5413/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3652/

File: core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java (hunk 5 above)
Comment (ShreelekhyaG) [reply core/src/main/java/org/apache/carbondata/core/indexstore/ExtendedBlocklet.java:236]: ok done

File: index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala [hunk 6]
```diff
@@ -86,8 +86,8 @@ class TestSIWithAddSegment extends QueryTest with BeforeAndAfterAll {
     sql(s"alter table maintable1 add segment options('path'='${ newSegmentPath }', " +
         s"'format'='carbon')")
     sql("CREATE INDEX maintable1_si  on table maintable1 (c) as 'carbondata'")
-    assert(sql("show segments for table maintable1_si").collect().length ==
-           sql("show segments for table maintable1").collect().length)
+    assert(sql("show segments for table maintable1_si").collect().length == 2)
+    assert(sql("show segments for table maintable1").collect().length == 3)
```
[the remark this answers is not in the document]
Comment (ShreelekhyaG) [reply index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:96]: Disabled SI table after alter add load and added a check to verify in test cases.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5414/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3653/

File: index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (Indhumathi27) [reply index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:96]: @akashrn5 why we need to disable SI ?

[the remark this answers is not in the document]
Comment (akashrn5) [reply index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:96]: since for external added segments we don't load SI, its scenario of segment mismatch of SI and main table, it may lead to wrong results or failure. We don't yet support of getting the results from SI for some segments and some segments from main table right.

[the remark this answers is not in the document]
Comment (Indhumathi27) [reply index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:96]: We use externalSegment Resolver for pruning external segments on main table and for remaining segments, we create a implicit filter to prune using SI. So, there will not be any wrong results for the query

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5420/

**Comment by Indhumathi27** [comment]
@ShreelekhyaG please update PR description

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3659/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3662/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5423/

File: index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala [hunk 7]
```diff
@@ -61,13 +61,16 @@ class TestSIWithAddSegment extends QueryTest with BeforeAndAfterAll {
   }
 
   test("test if the query hits SI after adding a segment to the main table") {
-    val d = sql("select * from maintable where c = 'm'")
-    assert(d.queryExecution.executedPlan.isInstanceOf[BroadCastSIFilterPushJoin])
+    val extSegmentQuery = sql("select * from maintable where c = 'm'")
+    val loadedSegmentQuery = sql("select * from maintable where c = 'k'")
```
Comment (Indhumathi27) [inline index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:65]: please verify result

Comment (ShreelekhyaG) [reply index/secondary-index/src/test/scala/org/apache/carbondata/spark/testsuite/secondaryindex/TestSIWithAddSegment.scala:65]: Done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5426/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3666/

**Comment by Indhumathi27** [comment]
LGTM

**Comment by akashrn5** [comment]
LGTM
